File: Pintpal_web/src/features/serving_game/data/servingGameScoreRepository.ts
/**
 * servingGameScoreRepository.ts
 *
 * Purpose: Persist Serving Rush personal bests and public serving_game_scores.
 * Connects to: ServingRushScreen / scoreboard; separate from pour_game_scores.
 */

import {
  collection,
  doc,
  getDocs,
  limit,
  orderBy,
  query,
  setDoc,
  where,
} from "firebase/firestore";
import { getFirebaseAuth, getFirebaseFirestore } from "@/utilities/firebase";
import {
  Collections,
  ServingGameScoreFields,
  UserFields,
} from "@/utilities/firebaseConstants";
import { loadUserDocument } from "@/features/auth/data/authRepository";

export type ServingScoreSubmission = {
  score: number;
  completedOrders: number;
  misses: number;
};

export type ServingScoreboardEntry = {
  id: string;
  userId: string;
  displayName: string;
  score: number;
  completedOrders: number;
  misses: number;
};

function isImproved(previous: number | null | undefined, candidate: number): boolean {
  return previous == null || candidate > previous;
}

export async function submitServingScore(
  submission: ServingScoreSubmission,
): Promise<{ personalBestUpdated: boolean; publicBoardUpdated: boolean }> {
  const auth = getFirebaseAuth();
  const user = auth.currentUser;
  if (!user) throw new Error("Not signed in");
  const db = getFirebaseFirestore();
  const docUser = await loadUserDocument(user.uid);
  if (!docUser) throw new Error("User profile missing");

  const previous = (docUser as { servingGameBestScore?: number | null })
    .servingGameBestScore;
  const personalBestUpdated = isImproved(previous, submission.score);
  if (personalBestUpdated) {
    await setDoc(
      doc(db, Collections.users, user.uid),
      { [UserFields.servingGameBestScore]: submission.score },
      { merge: true },
    );
  }

  let publicBoardUpdated = false;
  const optIn =
    (docUser as { servingGameScoreboardOptIn?: boolean }).servingGameScoreboardOptIn ===
    true;
  if (optIn && personalBestUpdated) {
    await setDoc(
      doc(db, Collections.servingGameScores, `${user.uid}_best`),
      {
        [ServingGameScoreFields.userId]: user.uid,
        [ServingGameScoreFields.displayName]:
          docUser.name?.trim() || user.email || "Player",
        [ServingGameScoreFields.score]: submission.score,
        [ServingGameScoreFields.completedOrders]: submission.completedOrders,
        [ServingGameScoreFields.misses]: submission.misses,
        [ServingGameScoreFields.updatedAt]: Date.now(),
        [ServingGameScoreFields.isPublic]: true,
      },
      { merge: true },
    );
    publicBoardUpdated = true;
  }

  return { personalBestUpdated, publicBoardUpdated };
}

export async function setServingScoreboardOptIn(optIn: boolean): Promise<void> {
  const auth = getFirebaseAuth();
  const user = auth.currentUser;
  if (!user) throw new Error("Not signed in");
  const db = getFirebaseFirestore();
  await setDoc(
    doc(db, Collections.users, user.uid),
    { [UserFields.servingGameScoreboardOptIn]: optIn },
    { merge: true },
  );
  if (!optIn) {
    await setDoc(
      doc(db, Collections.servingGameScores, `${user.uid}_best`),
      { [ServingGameScoreFields.isPublic]: false },
      { merge: true },
    );
  }
}

export async function loadPublicServingScoreboard(): Promise<
  ServingScoreboardEntry[]
> {
  const db = getFirebaseFirestore();
  const q = query(
    collection(db, Collections.servingGameScores),
    where(ServingGameScoreFields.isPublic, "==", true),
    orderBy(ServingGameScoreFields.score, "desc"),
    limit(50),
  );
  const snap = await getDocs(q);
  return snap.docs.map((d) => {
    const data = d.data();
    return {
      id: d.id,
      userId: String(data[ServingGameScoreFields.userId] ?? ""),
      displayName: String(data[ServingGameScoreFields.displayName] ?? "Player"),
      score: Number(data[ServingGameScoreFields.score] ?? 0),
      completedOrders: Number(data[ServingGameScoreFields.completedOrders] ?? 0),
      misses: Number(data[ServingGameScoreFields.misses] ?? 0),
    };
  });
}

File: pintpal_app/app/src/main/java/com/jdscott/pintpal/features/serving_game/data/ServingGameScoreRepository.kt
/**
 * ServingGameScoreRepository.kt
 *
 * Purpose: Persist Serving Rush personal bests + public serving_game_scores.
 */
package com.jdscott.pintpal.features.serving_game.data

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.jdscott.pintpal.features.pour_game.domain.PersonalBest
import com.jdscott.pintpal.utilities.FirebaseConstants
import kotlinx.coroutines.tasks.await

data class ServingSubmitResult(
    val personalBestUpdated: Boolean,
    val publicBoardUpdated: Boolean,
)

class ServingGameScoreRepository(
    private val auth: FirebaseAuth = FirebaseAuth.getInstance(),
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance(),
) {
    suspend fun submitScore(
        score: Int,
        completedOrders: Int,
        misses: Int,
    ): Result<ServingSubmitResult> = runCatching {
        val user = auth.currentUser ?: error("Not signed in")
        val userRef = firestore.collection(FirebaseConstants.Collections.USERS).document(user.uid)
        val snap = userRef.get().await()
        val data = snap.data.orEmpty()
        val previous = (data[FirebaseConstants.UserFields.SERVING_GAME_BEST] as? Number)?.toDouble()
        val personalBestUpdated = PersonalBest.isImproved(previous, score.toDouble())
        if (personalBestUpdated) {
            userRef.set(
                mapOf(FirebaseConstants.UserFields.SERVING_GAME_BEST to score),
                SetOptions.merge(),
            ).await()
        }
        val optIn =
            data[FirebaseConstants.UserFields.SERVING_GAME_SCOREBOARD_OPT_IN] as? Boolean ?: false
        var publicBoardUpdated = false
        if (optIn && personalBestUpdated) {
            val name = (data[FirebaseConstants.UserFields.NAME] as? String)?.trim()
                .orEmpty()
                .ifBlank { user.email.orEmpty().ifBlank { "Player" } }
            firestore.collection(FirebaseConstants.Collections.SERVING_GAME_SCORES)
                .document("${user.uid}_best")
                .set(
                    mapOf(
                        FirebaseConstants.ServingGameScoreFields.USER_ID to user.uid,
                        FirebaseConstants.ServingGameScoreFields.DISPLAY_NAME to name,
                        FirebaseConstants.ServingGameScoreFields.SCORE to score,
                        FirebaseConstants.ServingGameScoreFields.COMPLETED_ORDERS to completedOrders,
                        FirebaseConstants.ServingGameScoreFields.MISSES to misses,
                        FirebaseConstants.ServingGameScoreFields.UPDATED_AT to System.currentTimeMillis(),
                        FirebaseConstants.ServingGameScoreFields.IS_PUBLIC to true,
                    ),
                    SetOptions.merge(),
                )
                .await()
            publicBoardUpdated = true
        }
        ServingSubmitResult(personalBestUpdated, publicBoardUpdated)
    }

    suspend fun setScoreboardOptIn(optIn: Boolean): Result<Unit> = runCatching {
        val user = auth.currentUser ?: error("Not signed in")
        val userRef = firestore.collection(FirebaseConstants.Collections.USERS).document(user.uid)
        userRef.set(
            mapOf(FirebaseConstants.UserFields.SERVING_GAME_SCOREBOARD_OPT_IN to optIn),
            SetOptions.merge(),
        ).await()
        if (!optIn) {
            firestore.collection(FirebaseConstants.Collections.SERVING_GAME_SCORES)
                .document("${user.uid}_best")
                .delete()
                .await()
        }
    }

    suspend fun isScoreboardOptIn(): Boolean {
        val user = auth.currentUser ?: return false
        val snap = firestore.collection(FirebaseConstants.Collections.USERS)
            .document(user.uid)
            .get()
            .await()
        return snap.getBoolean(FirebaseConstants.UserFields.SERVING_GAME_SCOREBOARD_OPT_IN) == true
    }
}
